docs/tutorials/on-device-training.md

```diff
@@ -0,0 +1,1406 @@
+
+- Python development environment to prepare the training artifacts
+- Android Studio 4.1+
+- Android SDK 29+
+- Android NDK r21+
+- An Android device with a camera in [developer mode](https://developer.android.com/studio/debug/dev-options) with USB debugging enabled
+
+> **Note** The purpose of this tutorial is to show how ONNX Runtime can be used for On-Device Training. The author of this tutorial is not an expert in Android app development. The information provided in this tutorial is for educational purposes only and should not be considered as professional advice.
+
+> **Note** The entire android application is also made available on the [`onnxruntime-training-examples`](https://github.com/microsoft/onnxruntime-training-examples/tree/master/on_device_training/mobile/android/c-cpp) GitHub repository.
+
+## Offline Phase - Building the training artifacts
+
+1. Export the model to ONNX.
+
+    We start with a Pytorch model that has been pre-trained and export it to ONNX. The `MobileNetV2` model has been pretrained on the imagenet dataset that has data in 1000 categories. For our task of image classification, we want to only classify images in 4 classes. So, we change the last layer of the model to output 4 logits instead of 1,000.
+
+    ```python
+   import torch
+   import torchvision
+
+   model = torchvision.models.mobilenet_v2(
+      weights=torchvision.models.MobileNet_V2_Weights.IMAGENET1K_V2)
+
+   # The original model is trained on imagenet which has 1000 classes.
+   # For our image classification scenario, we need to classify among 4 categories.
+   # So we need to change the last layer of the model to have 4 outputs.
+   model.classifier[1] = torch.nn.Linear(1280, 4)
+
+   # Export the model to ONNX.
+   model_name = "mobilenetv2"
+   torch.onnx.export(model, torch.randn(1, 3, 224, 224),
+                     f"training_artifacts/{model_name}.onnx",
+                     input_names=["input"], output_names=["output"],
+                     dynamic_axes={"input": {0: "batch"}, "output": {0: "batch"}})
+    ```
+
+2. Define the trainable and non trainable parameters
+
+    ```python
+   import onnx
+
+   # Load the onnx model.
+   onnx_model = onnx.load(f"training_artifacts/{model_name}.onnx")
+
+   # Define the parameters that require their gradients to be computed
+   # and those that do not.
+   requires_grad = ["classifier.1.weight", "classifier.1.bias"]
+   frozen_params = [
+      param.name
+      for param in onnx_model.graph.initializer
+      if param.name not in requires_grad
+   ]
+    ```
+
+3. Generate the training artifacts. We will use the `CrossEntropyLoss` loss and the `AdamW` optimizer for this tutorial.
+
+    ```python
+   from onnxruntime.training import artifacts
```

Comment:
Is this a new library you've created for LTE? Some documentation on `artifacts` would be helpful.